jep/0000/README.adoc

```diff
@@ -0,0 +1,298 @@
+<j:jelly xmlns:j="jelly:core" xmlns:l="/lib/layout">
+  <l:view contentType="text/html;charset=UTF-8">
+    <html>
+      <!-- fancy SPA -->
+    </html>
+  </l:view>
+</j:jelly>
+----
+
+Or using Groovy:
+
+[source,groovy]
+----
+def l = namespace('/lib/layout')
+l.view(contentType: 'text/html;charset=UTF-8') {
+  html {
+    // ...
+  }
+}
+----
+
+=== Model Class Annotations
+
+A model class can be annotated with `@StaplerViews` to list out which views are dispatchable.
+Likewise, using `@StaplerFragments`, the list of views considered fragments are declared.
+Any views not listed in those annotations on the model class or any of its parents remain unaffected by this filtering.
+Use of these annotations takes precedence over the presence of the layout tags listed in the above subsection.
+
+This provides methods to specify and override default views and fragments.
+For example, suppose Jenkins defines an abstract class `BaseClass` which has views `index`, `configure`, `create`, and `delete`.
+Implementation classes are expected to provide their own fragment view `config` which gets included in `configure` and `create`.
+This can be encoded explicitly in the model class:
+
+[source,java]
+----
+@StaplerViews({"configure", "create", "delete"}) // index is implicit
+@StaplerFragments("config")
+public abstract class BaseClass {
+    ...
+}
+----
+
+By adding the annotations to Jenkins, implementation classes (typically from plugins) can be filtered explicitly via the parent annotations without updating the downstream code.
+Implementation classes may further define additional views or fragments.
+For example, suppose `ExtendedClass` extends `BaseClass` and adds its own optional fragment view `advanced` which its implementation classes can include.
+
+[source,java]
+----
+@StaplerFragments("advanced")
+public class ExtendedClass extends BaseClass {
+    ...
+}
+----
+
+=== Whitelist File
+
+Jenkins ships with a default whitelist of common, known views in plugins that would otherwise be broken by the above rules.
+This file consists of class names and their respective views and fragments.
+These entries have the same effect as annotating the listed classes with the above annotations.
+The syntax for this file is defined below.
+
+[source,bnf]
+----
+<line> ::= <fully-qualified-model-class-name> <ws> <view-list>
+<view-list> ::= <view-entry> | <view-entry> <ws> <view-list>
+<view-entry> ::= <view-name> | "!" <view-name>
+<ws> ::= " " | " " <ws>
+----
+
+For each view entry, if the view name is prefixed by a `!`, then it is added as a fragment view.
+Otherwise, the view name is added as a dispatchable view.
+
+Additional whitelist entries can be specified by the file `$JENKINS_HOME/stapler-views-whitelist.txt` or whatever path is specified by the system property `jenkins.security.stapler.StaplerDispatchValidator.whitelist`.
+
+== Motivation
+
+During exploratory testing, the Jenkins security team discovered a class of vulnerabilities in Jenkins.
```

Comment:
Strictly speaking, it's just one (with _many_ instances) -- information disclosure.

SECURITY-595 was different, as it allowed a variety of different outcomes.